### DISCUSSION TITLE
I want to generate thumbnails of multiple elements together, but an exception occurred

### DISCUSSION THREAD

**Discussion by w934423231** [fabricjs/fabric.js#8239] [Q&A] [answered]
This is my test case. Although part of it is simulated, there is still an error in my real environment. I can't solve it
https://codepen.io/w934423231/pen/ExLxPxW?editors=1111

**Comment by w934423231** [comment]
![image](https://user-images.githubusercontent.com/38958128/188045502-371335c5-9791-4890-9250-958c56bf961f.png)
![image](https://user-images.githubusercontent.com/38958128/188045555-e4bcd9f7-2390-4b25-80ae-138673492343.png)
Through checking, only after calling this line of code, such an error will occur, because the canvas of the element is set to undefind

**Reply by w934423231** [reply]
This is the effect before I call newgroup. Todataurl()
![9J)GB5 8G$D1P)X{4I~R_LQ](https://user-images.githubusercontent.com/38958128/188045808-8ee29a20-0b6a-4b65-ba96-cc0d644c34c9.png)
But after I called it, it became undefind
![QLFTYIRFR3 EGWUI3 MP7SR](https://user-images.githubusercontent.com/38958128/188045878-8490d7be-c2ba-4494-9ea0-240dfa200103.png)

**Comment by ShaMan123** [comment, marked as answer]
You are right
Objects need a ref to canvas
The problem is your object tree is not strict meaning an object has more than one parent.
that is the bug.
v6 doesn't allow that

**Reply by w934423231** [reply]
Please know how I can complete this function more reasonably

**Reply by ShaMan123** [reply]
You add objects to canvas
Then you add them to a group.
Then you ungroup.
Why?
Don't put them in the group.
You can do canvas.toDataURL.
If you need the grouping logic (which is bad under v6) then understand that the objects loose the ref to canvas after ungroup. So handle that.

**Reply by w934423231** [reply]
Because the current system does not handle any grouping function, if I group randomly, my program will run abnormally
